src/bin/pg_autoctl/pghba.c

```diff
@@ -155,15 +202,176 @@ pghba_ensure_host_rule_exists(const char *hbaFilePath,
 		return false;
 	}
 
-	/* write the new postgresql.conf */
+	/* write the new pg_hba.conf */
+	if (!write_file(newHbaContents->data, newHbaContents->len, hbaFilePath))
+	{
+		/* write_file logs an error */
+		destroyPQExpBuffer(newHbaContents);
+		return false;
+	}
+
+	destroyPQExpBuffer(newHbaContents);
+
+	log_debug("Wrote new %s", hbaFilePath);
+
+	return true;
+}
+
+
+/*
+ * pghba_ensure_host_rules_exist ensures that we have all the rules needed for
+ * the given array of nodes, as retrived from the monitor for our formation and
+ * group, presumably.
+ *
+ * Each node in the array needs two rules:
+ *
+ *  host(ssl) replication "pgautofailover_replicator" nodename/ip trust
+ *  host(ssl) "dbname"    "pgautofailover_replicator" nodename/ip trust
+ */
+bool
+pghba_ensure_host_rules_exist(const char *hbaFilePath,
+							  NodeAddressArray *nodesArray,
+							  bool ssl,
+							  const char *database,
+							  const char *username,
+							  const char *authenticationScheme)
+{
+	PQExpBuffer newHbaContents = createPQExpBuffer();
+	char *currentHbaContents = NULL;
+	long currentHbaSize = 0L;
+	char *includeLine = NULL;
+
+	int nodeIndex = 0;
+
+	if (newHbaContents == NULL)
+	{
+		log_error("Failed to allocate memory");
+		return false;
+	}
+
+	if (!read_file(hbaFilePath, &currentHbaContents, &currentHbaSize))
+	{
+		/* read_file logs an error */
+		return false;
+	}
+
+	/* always begin with the existing HBA file */
+	appendPQExpBufferStr(newHbaContents, currentHbaContents);
+
+	for (nodeIndex = 0; nodeIndex < nodesArray->count; nodeIndex++)
+	{
+		NodeAddress *node = &(nodesArray->nodes[nodeIndex]);
+		PQExpBuffer hbaLineReplicationBuffer = createPQExpBuffer();
+		PQExpBuffer hbaLineDatabaseBuffer = createPQExpBuffer();
+
+		int hbaLinesIndex = 0;
+		PQExpBuffer hbaLines[] = {
+			hbaLineReplicationBuffer,
+			hbaLineDatabaseBuffer,
+			NULL
+		};
+
+		if (hbaLineReplicationBuffer == NULL ||
+			hbaLineDatabaseBuffer == NULL)
+		{
+			log_error("Failed to allocate memory");
+			return false;
```

Comment:
On this `return false` and the ones below, we leak `hbaLineReplicationBuffer`, `hbaLineDatabaseBuffer` and `newHbaContents`

Reply:
Fixed, thanks.